### PR TITLE
Feature/tags

### PR DESCRIPTION
**Pull request by weeryan17**
# Pull request

 - [x] I have read and agreed to the [code of conduct](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CODE_OF_CONDUCT.md).
 - [x] I have read and complied with the [contributing guidelines](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CONTRIBUTING.md).
 - [x] What I'm implementing was assigned to me or was not being worked on and is a somewhat requested feature. For reference see our [GitHub projects](https://github.com/orgs/CascadeBot/projects).
 - [x] I have tested all of my changes.
 
## Added/Changed
Added tags, placeholders for said tags, and associated commands

I still need to do these
 - [x] Find a nice way to handle categories for tags
 - [x] Implement command logic for tags
 - [x] Add tag edit functionality
 - [x] A way to list placeholders

### CODE REVIEW

File: src/main/java/org/cascadebot/cascadebot/events/CommandListener.java [hunk 1]
```diff
@@ -185,11 +185,13 @@ private void processCommands(GuildMessageReceivedEvent event, GuildData guildDat
             dispatchCommand(cmd, context);
         }
 
-        //TODO maybe add setting for this?
-        if (guildData.getTagInfo().containsKey(trigger)) {
-            Tag tag = guildData.getTag(trigger);
+        if(guildData.getSettings().willAllowTagCommands()) {
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/events/CommandListener.java]: Format

File: src/main/java/org/cascadebot/cascadebot/events/CommandListener.java [hunk 2]
```diff
@@ -185,11 +185,13 @@ private void processCommands(GuildMessageReceivedEvent event, GuildData guildDat
...
+            if (guildData.getTagInfo().containsKey(trigger)) {
+                Tag tag = guildData.getTag(trigger);
 
-            context.reply(tag.formatTag(context)); //Are we adding perms to tags? If so i need to implement that. - Ryan
+                context.reply(tag.formatTag(context)); //Are we adding perms to tags? If so i need to implement that. - Ryan
+                CascadeBot.LOGGER.info("Tag {} executed by {} with args {}", trigger, context.getUser().getAsTag(), Arrays.toString(context.getArgs()));
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/events/CommandListener.java]: I'm honestly tempted to say that we should leave this at debug level

File: src/main/java/org/cascadebot/cascadebot/commands/subcommands/tag/TagCreateSubCommand.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+public class TagCreateSubCommand implements ICommandExecutable {
+
+    @Override
+    public void onCommand(Member sender, CommandContext context) {
+        if (context.getArgs().length < 2) {
+            context.getUIMessaging().replyUsage(this, "tag");
+            return;
+        }
+        context.getData().addTag(context.getArg(0), new Tag(context.getMessage(1), "tag")); //TODO handle category
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/commands/subcommands/tag/TagCreateSubCommand.java]: `//TODO handle category`

File: src/main/java/org/cascadebot/cascadebot/commands/subcommands/tag/TagDeleteSubCommand.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import org.cascadebot.cascadebot.commandmeta.ICommandExecutable;
+import org.cascadebot.cascadebot.permissions.CascadePermission;
+
+import java.util.Set;
+
+public class TagDeleteSubCommand implements ICommandExecutable {
+
+    @Override
+    public void onCommand(Member sender, CommandContext context) {
+        if (context.getData().removeTag(context.getArg(0))) {
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/commands/subcommands/tag/TagDeleteSubCommand.java]: Check args length

File: src/main/java/org/cascadebot/cascadebot/data/objects/GuildData.java [hunk 5]
```diff
@@ -22,11 +22,7 @@
 import org.cascadebot.cascadebot.utils.buttons.ButtonsCache;
 import org.cascadebot.cascadebot.utils.pagination.PageCache;
 
-import java.util.Collection;
-import java.util.Collections;
-import java.util.Date;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/data/objects/GuildData.java]: * import

File: src/main/java/org/cascadebot/cascadebot/events/CommandListener.java [hunk 6]
```diff
@@ -183,6 +184,15 @@ private void processCommands(GuildMessageReceivedEvent event, GuildData guildDat
             }
             dispatchCommand(cmd, context);
         }
+
+        if (guildData.getSettings().willAllowTagCommands()) {
+            if (guildData.getTagInfo().containsKey(trigger)) {
+                Tag tag = guildData.getTag(trigger);
+
+                context.reply(tag.formatTag(context)); //Are we adding perms to tags? If so i need to implement that. - Ryan
```
Comment (binaryoverload) [inline src/main/java/org/cascadebot/cascadebot/events/CommandListener.java]: Yes we are adding perms to tags

File: src/main/java/org/cascadebot/cascadebot/events/CommandListener.java (hunk 2 above)
Comment (binaryoverload) [reply src/main/java/org/cascadebot/cascadebot/events/CommandListener.java]: Bump

File: src/main/java/org/cascadebot/cascadebot/commands/subcommands/tag/TagCreateSubCommand.java (hunk 3 above)
Comment (weeryan17) [reply src/main/java/org/cascadebot/cascadebot/commands/subcommands/tag/TagCreateSubCommand.java]: Whops forgot to remove that

File: src/main/java/org/cascadebot/cascadebot/events/CommandListener.java (hunk 2 above)
Comment (weeryan17) [reply src/main/java/org/cascadebot/cascadebot/events/CommandListener.java]: I doubt tags will be used more then commands and we're already logging all commands used.

File: src/main/java/org/cascadebot/cascadebot/events/CommandListener.java (hunk 6 above)
Comment (weeryan17) [reply src/main/java/org/cascadebot/cascadebot/events/CommandListener.java]: Think we should hold of this pr till we get per guild permissions or just merge it now?

Comment (binaryoverload) [reply src/main/java/org/cascadebot/cascadebot/events/CommandListener.java]: Make this comment an actual TODO so we can find it and merge it I think

**Pull request merged**